code/game/turfs/simulated/river.dm

```diff
@@ -7,11 +7,15 @@
 /proc/spawn_rivers(target_z, nodes = 4, turf_type = /turf/open/lava/smooth/lava_land_surface, whitelist_area = /area/lavaland/surface/outdoors, min_x = RANDOM_LOWER_X, min_y = RANDOM_LOWER_Y, max_x = RANDOM_UPPER_X, max_y = RANDOM_UPPER_Y)
 	var/list/river_nodes = list()
 	var/num_spawned = 0
-	while(num_spawned < nodes)
-		var/turf/F = locate(rand(min_x, max_x), rand(min_y, max_y), target_z)
-
-		river_nodes += new /obj/effect/landmark/river_waypoint(F)
-		num_spawned++
+	var/list/possible_locs = block(locate(min_x, min_y, target_z), locate(max_x, max_y, target_z))
+	while(num_spawned < nodes && possible_locs.len)
+		var/turf/T = pick(possible_locs)
+		var/area/A = get_area(T)
+		if(!istype(A, whitelist_area) || (T.flags_1 & NO_LAVA_GEN_1))
```

Comment:
This part fixes  #27719, it would previously just leave one single lava tile if the river tried to start in a non-whitelisted area.